src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs

```diff
@@ -207,52 +238,209 @@ public static FormulaValue Text(EvalVisitor runner, EvalVisitorContext context,
 
+        internal static string ExpandDateTimeFormatSpecifiers(string format, CultureInfo culture)
+        {
             var info = DateTimeFormatInfo.GetInstance(culture);
 
-            switch (format.ToLower().Trim('\''))
+            switch (format.ToLower())
             {
-                case "shortdatetime24":
+                case "'shortdatetime24'":
                     // TODO: This might be wrong for some cultures
                     return ReplaceWith24HourClock(info.ShortDatePattern + " " + info.ShortTimePattern);
-                case "shortdatetime":
+                case "'shortdatetime'":
                     // TODO: This might be wrong for some cultures
                     return info.ShortDatePattern + " " + info.ShortTimePattern;
-                case "shorttime24":
+                case "'shorttime24'":
                     return ReplaceWith24HourClock(info.ShortTimePattern);
-                case "shorttime":
+                case "'shorttime'":
                     return info.ShortTimePattern;
-                case "shortdate":
+                case "'shortdate'":
                     return info.ShortDatePattern;
-                case "longdatetime24":
+                case "'longdatetime24'":
                     return ReplaceWith24HourClock(info.FullDateTimePattern);
-                case "longdatetime":
+                case "'longdatetime'":
                     return info.FullDateTimePattern;
-                case "longtime24":
+                case "'longtime24'":
                     return ReplaceWith24HourClock(info.LongTimePattern);
-                case "longtime":
+                case "'longtime'":
                     return info.LongTimePattern;
-                case "longdate":
+                case "'longdate'":
                     return info.LongDatePattern;
-                case "utc":
+                case "'utc'":
                     return info.UniversalSortableDateTimePattern;
+                default:
+                    return format;
             }
+        }
+
+        private static string ReplaceWith24HourClock(string format)
+        {
+            format = Regex.Replace(format, "[hH]", "H");
+            format = Regex.Replace(format, "t+", "");
+
+            return format.Trim();
+        }
+
+        private static string ResolveDateTimeFormatAmbiguities(EvalVisitor runner, string format, DateTime dateTime, CultureInfo culture)
+        {
+            format = ReplaceDoubleQuotedStrings(runner, format, out var replaceList);        
+            format = TokenizeDatetimeFormat(runner, format);
+            format = DetokenizeDatetimeFormat(format, dateTime, culture);
+            format = RestoreDoubleQuotedStrings(runner, format, replaceList);
 
             return format;
         }
 
-        private static string ReplaceWith24HourClock(string format)
+        private static string RestoreDoubleQuotedStrings(EvalVisitor runner, string format, List<string> repleaceList)
```

Comment:
>EvalVisitor runner, [](http://example.com/codeflow?start=57&length=19)

Least privilege - do we need to pass EvalVisitor (a large powerful object); or can we just pass the CancellationToken?

Reply:
CancellationToken was not available from EvalVisitor. Now it is available as a field.